crates/snapbox/src/data.rs

```diff
@@ -155,22 +176,52 @@ impl Data {
         match &self.inner {
             DataInner::Binary(_) => None,
             DataInner::Text(data) => Some(data.as_str()),
+            #[cfg(feature = "json")]
+            DataInner::Json(json) => serde_json::to_string(json).map(|j| &j).ok(),
         }
     }
 
     pub fn as_bytes(&self) -> &[u8] {
         match &self.inner {
             DataInner::Binary(data) => data,
             DataInner::Text(data) => data.as_bytes(),
+            #[cfg(feature = "json")]
+            DataInner::Json(json) => &serde_json::to_vec(json).unwrap(),
         }
     }
+
+    #[cfg(feature = "detect-extension")]
+    fn deserialize(
+        data: String,
+        extension: Option<&std::ffi::OsStr>,
+    ) -> Result<Self, crate::Error> {
+        if let Some(ext) = extension {
+            match ext.to_str() {
+                #[cfg(feature = "json")]
+                Some("json") => Ok(Self::json(serde_json::from_str(&data)?)),
+                _ => Ok(Self::text(data)),
+            }
+        } else {
+            Ok(Self::text(data))
+        }
+    }
+
+    #[cfg(not(feature = "detect-extension"))]
+    fn deserialize(
+        data: String,
+        _extension: Option<&std::ffi::OsStr>,
+    ) -> Result<Self, crate::Error> {
+        Ok(Self::text(data))
+    }
 }
 
 impl std::fmt::Display for Data {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match &self.inner {
             DataInner::Binary(data) => String::from_utf8_lossy(data).fmt(f),
             DataInner::Text(data) => data.fmt(f),
+            #[cfg(feature = "json")]
+            DataInner::Json(json) => json.fmt(f),
```

Comment:
Should this be pretty or regular?

Reply:
More than likely pretty. I will have to figure out how to achieve that